### PR TITLE
Vm pertb est

### PR DESCRIPTION
**Pull request by shot0829**
#### Description
doc for estimation numbers:
https://wiki.canterbury.ac.nz/display/QuakeCore/VM+pertbation+estimation
#### Checklist
- Have you updated the README (yes/no)?
- Have you updated the CHANGELOG (yes/no)?

### CODE REVIEW

**Review by sungeunbae — APPROVED** [review]

File: estimation/estimate_wct.py [hunk 1]
```diff
@@ -390,6 +388,37 @@ def estimate_BB_chours(
...
+        "b": 5.917_060_637_004_801e-20,
+        "c": 0.130_000_000_000_000_4,
+    }
+
+    vm_size = data[:, 0]
+
+    core_hours = (
+        (vm_size * coefficients["a"])
+        + ((vm_size ** 2) * coefficients["b"])
+        + coefficients["c"]
```
[the remark this answers is not in the document]
Comment (shot0829) [reply estimation/estimate_wct.py]: re-ordered.

File: scripts/schedulers/slurm.py [hunk 2]
```diff
@@ -120,8 +120,27 @@ def cancel_job(self, job_id, target_machine=None):
...
+        scheduler_header_command_dict = {
+            "time": "--time={value}",
+            "job_name": "--job-name={value}",
+            "ncpus": "--cpus-per-task={value}",
+            "nodes": "--nodes={value}",
+            "ntasks": "--ntasks={value}",
+        }
+        scheduler_args_commands = ""
+        for key, value in scheduler_arguments.items():
+            if key in scheduler_arguments.keys():
```
Comment (jpa198) [inline scripts/schedulers/slurm.py:136]: Always returns true

Comment (shot0829) [reply scripts/schedulers/slurm.py]: haha, fixed

**Review by jasonmotha — DISMISSED** [review]
Looks pretty good.

File: estimation/estimate_wct.py [hunk 3]
```diff
@@ -390,6 +390,37 @@ def estimate_BB_chours(
...
+
+    core_hours, run_time = est_VM_PERT_chours(data)
+    return core_hours[0], run_time[0]
+
+
+def est_VM_PERT_chours(data: np.ndarray):
+
+    if data.shape[1] != 2:
+        raise Exception(
+            "Invalid input data, has to 3 columns. " "One for each feature."
```
Comment (jpa198) [inline estimation/estimate_wct.py:404]: "has to 3"?

Comment (shot0829) [reply estimation/estimate_wct.py]: fixed

File: scripts/submit_vm_pert.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+from qcore import simulation_structure as sim_struct
+from shared_workflow.shared import set_wct
+from shared_workflow.platform_config import (
+    HPC,
+    platform_config,
+    get_platform_specific_script,
+    get_target_machine,
+)
+
+DEFAULT_CPUS = 4
```
Comment (jpa198) [inline scripts/submit_vm_pert.py:24]: This should be in the platform configs

File: estimation/estimate_wct.py (hunk 3 above)
Comment (jasonmotha) [reply estimation/estimate_wct.py]: Since this is a copy pasted string - I think we should make it a constant.

Comment (jasonmotha) [reply estimation/estimate_wct.py]: Maybe use .format as we have different n_columns.

File: scripts/submit_vm_pert.py (hunk 4 above)
Comment (shot0829) [reply scripts/submit_vm_pert.py]: done.

File: estimation/estimate_wct.py (hunk 3 above)
Comment (shot0829) [reply estimation/estimate_wct.py]: good idea. done.

**Review by jasonmotha — APPROVED** [review]

File: estimation/estimate_wct.py [hunk 5]
```diff
@@ -25,6 +25,7 @@
 PHYSICAL_NCORES_PER_NODE = config.qconfig[config.ConfigKeys.cores_per_node.name]
 
 CH_SAFETY_FACTOR = 1.5
+ERROR_MSG_SHAPE_MISMATCH = "Invalid input data, has to be {required_column_count} columns. One for each feature."
```
Comment (jpa198) [inline estimation/estimate_wct.py:28]: Should we just make a custom exception with the same message, and you input the column count instead?
e.g.
class EstimationInputException(Exception):
    ERROR_MSG_SHAPE_MISMATCH = "..."
    def __init__(self, column_count, *args, **kwargs):
        super().__init__(ERRO_MSG_SHAPE_MISMATCH.format(required_column_count=column_count), *args, **kwargs)

Comment (jpa198) [reply estimation/estimate_wct.py:28]: Ugh, thanks for nuking the formatting github

Comment (shot0829) [reply estimation/estimate_wct.py:28]: that sounds like an overkill, at least for now. 
we have only this one kind of error at the moment, constructing our own handler seems not necessary.

Comment (shot0829) [reply estimation/estimate_wct.py:28]: if we have a lot of checking/error handling that might be more useful.

**Pull request merged**